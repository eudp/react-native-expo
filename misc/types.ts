export type FilterType = "all" | "reedemed" | "noReedemed";

export type ProductType = {
  createdAt?: string;
  product?: string;
  points?: number;
  image?: string;
  is_redemption?: boolean;
  id?: string;
};
