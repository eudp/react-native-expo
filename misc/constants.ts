import { FilterType } from "./types";

export const FILTERS = {
  ALL: "all" as FilterType,
  REEDEMED: "reedemed" as FilterType,
  NO_REEDEMED: "noReedemed" as FilterType,
};

export const API_URL =
  "https://6222994f666291106a29f999.mockapi.io/api/v1/products";

export const COLORS = {
  green: "#34C759",
  red: "#FF0000",
};
