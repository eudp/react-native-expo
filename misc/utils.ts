import { ProductType } from "./types";

export const getTotalPoints = (products: ProductType[]) =>
  products.reduce((acc, product) => {
    if (product.is_redemption) {
      return acc - product.points;
    }
    return acc + product.points;
  }, 0);
