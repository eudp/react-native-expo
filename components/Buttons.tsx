import { StyleSheet, View } from "react-native";
import Button from "./Button";
import { FilterType } from "../misc/types";
import { FILTERS } from "../misc/constants";

interface Props {
  handleFilter: (filter: string) => void;
  filter: FilterType;
}

const Buttons = ({ handleFilter, filter }: Props) => (
  <View style={styles.buttonContainer}>
    {filter === FILTERS.ALL ? (
      <>
        <Button
          text="Ganados"
          onPress={() => {
            handleFilter(FILTERS.NO_REEDEMED);
          }}
          styleButton={{ marginRight: 13 }}
        />
        <Button
          text="Canjeados"
          onPress={() => {
            handleFilter(FILTERS.REEDEMED);
          }}
        />
      </>
    ) : (
      <Button
        text="Todos"
        onPress={() => {
          handleFilter(FILTERS.ALL);
        }}
      />
    )}
  </View>
);

const styles = StyleSheet.create({
  buttonContainer: {
    flex: 2,
    flexDirection: "row",
    justifyContent: "center",
    alignItems: "flex-end",
  },
});

export default Buttons;
