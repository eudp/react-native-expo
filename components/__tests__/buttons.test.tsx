import React from "react";
import renderer from "react-test-renderer";
import Buttons from "../Buttons";
import Button from "../Button";
import { FILTERS } from "../../misc/constants";

describe("Buttons", () => {
  const handleFilter = jest.fn();

  it("renders correctly when filter is all", () => {
    const tree = renderer
      .create(<Buttons handleFilter={handleFilter} filter={FILTERS.ALL} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders correctly when filter is not all", () => {
    const tree = renderer
      .create(
        <Buttons handleFilter={handleFilter} filter={FILTERS.NO_REEDEMED} />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("calls handleFilter with correct filter when 'Ganados' button is pressed", () => {
    const component = renderer.create(
      <Buttons handleFilter={handleFilter} filter={FILTERS.ALL} />
    );

    const buttons = component.root.findAllByType(Button);
    buttons[0].props.onPress();
    expect(handleFilter).toHaveBeenCalledWith(FILTERS.NO_REEDEMED);
  });

  it("calls handleFilter with correct filter when 'Canjeados' button is pressed", () => {
    const component = renderer.create(
      <Buttons handleFilter={handleFilter} filter={FILTERS.ALL} />
    );

    const buttons = component.root.findAllByType(Button);
    buttons[1].props.onPress();
    expect(handleFilter).toHaveBeenCalledWith(FILTERS.REEDEMED);
  });

  it("calls handleFilter with correct filter when 'Todos' button is pressed", () => {
    const component = renderer.create(
      <Buttons handleFilter={handleFilter} filter={FILTERS.NO_REEDEMED} />
    );

    const buttons = component.root.findAllByType(Button);
    buttons[0].props.onPress();
    expect(handleFilter).toHaveBeenCalledWith(FILTERS.ALL);
  });
});
