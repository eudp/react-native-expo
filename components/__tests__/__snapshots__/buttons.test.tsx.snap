// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Buttons renders correctly when filter is all 1`] = `
<View
  style={
    {
      "alignItems": "flex-end",
      "flex": 2,
      "flexDirection": "row",
      "justifyContent": "center",
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#334FFA",
        "borderRadius": 10,
        "flex": 1,
        "marginRight": 13,
        "opacity": 1,
        "padding": 10,
        "paddingVertical": 20,
      }
    }
  >
    <Text
      style={
        [
          {
            "fontFamily": "Avenir-Regular",
          },
          {
            "color": "#FFFFFF",
            "fontFamily": "Avenir-Black",
          },
        ]
      }
    >
      Ganados
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#334FFA",
        "borderRadius": 10,
        "flex": 1,
        "opacity": 1,
        "padding": 10,
        "paddingVertical": 20,
      }
    }
  >
    <Text
      style={
        [
          {
            "fontFamily": "Avenir-Regular",
          },
          {
            "color": "#FFFFFF",
            "fontFamily": "Avenir-Black",
          },
        ]
      }
    >
      Canjeados
    </Text>
  </View>
</View>
`;

exports[`Buttons renders correctly when filter is not all 1`] = `
<View
  style={
    {
      "alignItems": "flex-end",
      "flex": 2,
      "flexDirection": "row",
      "justifyContent": "center",
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#334FFA",
        "borderRadius": 10,
        "flex": 1,
        "opacity": 1,
        "padding": 10,
        "paddingVertical": 20,
      }
    }
  >
    <Text
      style={
        [
          {
            "fontFamily": "Avenir-Regular",
          },
          {
            "color": "#FFFFFF",
            "fontFamily": "Avenir-Black",
          },
        ]
      }
    >
      Todos
    </Text>
  </View>
</View>
`;
